NitroxClient/Debuggers/SceneDebugger.cs

```diff
@@ -213,7 +222,73 @@ private void RenderTabHierarchy()
+                {
+                    if (searchTaskCancelTokenSource.IsCancellationRequested)
+                    {
+                        currentSearchItem = string.Empty;
+                        return false;
+                    }
+
+                    currentSearchItem = go.name;
+                    // Search the name of the selected gameobject
+                    return go.name.IndexOf(searchText, StringComparison.OrdinalIgnoreCase) > -1 ||
+                    go.GetComponents<Component>().Any(c => c.GetType().Name.IndexOf(searchText, StringComparison.OrdinalIgnoreCase) > -1) ||                        // Search the name of the child gameobjects.
+                    go.GetComponentsInChildren<Transform>().Any(t => t.gameObject != go && filter(t.gameObject, searchText));
+                }
+
+
+                // Searchbar
+                using (new GUILayout.VerticalScope("box"))
+                {
+                    using (new GUILayout.HorizontalScope())
+                    {                        
+                        searchText = GUILayout.TextField(searchText);
+
+                        // This button is used for both clearing the text and
+                        // cancelling any running search.
+                        if (GUILayout.Button("X", GUILayout.Width(30)))
+                        {
+                            searchText = string.Empty;
+                            searchTaskCancelTokenSource.Cancel();
+                        }
+
+                        // If there's no search text, then show all items
+                        if (string.IsNullOrEmpty(searchText))
+                        {
+                            filteredObjects = showObjects;
+                        }
+                    }
+
+                    // Disable the search button while a search is running
+                    GUI.enabled = !isSearching;
+                    // If we click the search bar, filter down the list
+                    if (GUILayout.Button("Search") || shouldSearchAgain)
+                    {
+                        shouldSearchAgain = false;
+                        isSearching = true;
+                        searchTaskCancelTokenSource = new CancellationTokenSource();
+                        Task.Run(() =>
+                        {
+                            filteredObjects = showObjects.Where(go => filter(go, searchText)).ToList();
+                            isSearching = false;
+                            currentSearchItem = string.Empty;
+                        }, searchTaskCancelTokenSource.Token);
+                    }
+                    // Make sure to re-enable the GUI for everyone else
+                    GUI.enabled = true;
+
+                    // Display some sort of indication to the user that we haven't just frozen up
+                    GUILayout.Label("Current Search Item: " + currentSearchItem);                    
+                }
```

Comment:
The hierarchy tab is getting complicated. Could you refactor the code to use a new inner class (to the SceneDebugger) that handles the filtering? Similarly to the `IDrawer` system but just for the search bar.